subsys/net/lib/sockets/sockets_packet.c

```diff
@@ -0,0 +1,339 @@
+{
+	struct k_poll_event events[] = {
+		K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_FIFO_DATA_AVAILABLE,
+					 K_POLL_MODE_NOTIFY_ONLY, fifo),
+	};
+
+	return k_poll(events, ARRAY_SIZE(events), timeout);
+}
+
+int zpacket_socket(int family, int type, int proto)
+{
+	struct net_context *ctx;
+	int fd;
+	int ret;
+
+	if (type != SOCK_RAW || proto != ETH_P_ALL) {
+		errno = -EOPNOTSUPP;
+		return -1;
+	}
+
+	fd = z_reserve_fd();
+	if (fd < 0) {
+		return -1;
+	}
+
+	ret = net_context_get(family, type, proto, &ctx);
+	if (ret < 0) {
+		z_free_fd(fd);
+		errno = -ret;
+		return -1;
+	}
+
+	/* Initialize user_data, all other calls will preserve it */
+	ctx->user_data = NULL;
+
+	/* recv_q and accept_q are in union */
+	k_fifo_init(&ctx->recv_q);
+
+#ifdef CONFIG_USERSPACE
+	/* Set net context object as initialized and grant access to the
+	 * calling thread (and only the calling thread)
+	 */
+	_k_object_recycle(ctx);
+#endif
+
+	z_finalize_fd(fd, ctx,
+		      (const struct fd_op_vtable *)&packet_sock_fd_op_vtable);
+
+	return fd;
+}
+
+static void zpacket_received_cb(struct net_context *ctx,
+				struct net_pkt *pkt,
+				union net_ip_header *ip_hdr,
+				union net_proto_header *proto_hdr,
+				int status,
+				void *user_data)
+{
+	NET_DBG("ctx=%p, pkt=%p, st=%d, user_data=%p", ctx, pkt, status,
+		user_data);
+
+	/* if pkt is NULL, EOF */
+	if (!pkt) {
+		struct net_pkt *last_pkt = k_fifo_peek_tail(&ctx->recv_q);
+
+		if (!last_pkt) {
+			/* If there're no packets in the queue, recv() may
+			 * be blocked waiting on it to become non-empty,
+			 * so cancel that wait.
+			 */
+			sock_set_eof(ctx);
+			k_fifo_cancel_wait(&ctx->recv_q);
+			NET_DBG("Marked socket %p as peer-closed", ctx);
+		} else {
+			net_pkt_set_eof(last_pkt, true);
+			NET_DBG("Set EOF flag on pkt %p", ctx);
+		}
+
+		return;
+	}
+
+	/* Normal packet */
+	net_pkt_set_eof(pkt, false);
+
+	k_fifo_put(&ctx->recv_q, pkt);
+}
+
+static int zpacket_bind_ctx(struct net_context *ctx,
+			    const struct sockaddr *addr,
+			    socklen_t addrlen)
+{
+	int ret = 0;
+
+	ret = net_context_bind(ctx, addr, addrlen);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	/* For packet socket, we expect to receive packets after call
+	 * to bind().
+	 */
+	ret = net_context_recv(ctx, zpacket_received_cb, K_NO_WAIT,
+			       ctx->user_data);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	return 0;
+}
+
+ssize_t zpacket_sendto_ctx(struct net_context *ctx, const void *buf, size_t len,
+			   int flags, const struct sockaddr *dest_addr,
+			   socklen_t addrlen)
+{
+	s32_t timeout = K_FOREVER;
+	int status;
+
+	if (!dest_addr) {
+		errno = EDESTADDRREQ;
+		return -1;
+	}
+
+	if ((flags & ZSOCK_MSG_DONTWAIT) || sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
+	}
+
+	/* Register the callback before sending in order to receive the response
+	 * from the peer.
+	 */
+
+	status = net_context_recv(ctx, zpacket_received_cb, K_NO_WAIT,
```

Comment:
Why do we need to register the zpacket_received_cb here?

Reply:
That is part of the packet socket patch that I had to build on top of. So that is discussed on #12564. I copied the same code to sockets_can.c but it might be that we do not need this call after all. That needs more investigation.